### PR TITLE
Color redesign & annotations fixes

### PR DESCRIPTION
**Pull request by macdrevx**
## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Update the changelog
 - [x] Update the migration guide, API Docs, Markdown files - Readme, Developing, etc

#### Breaking Changes

- Renamed `ColorRepresentable` to `StyleColor`
- Removed the argument label from `StyleColor`'s `UIColor` initializer
- Renamed `ColorRepresentable.rgbaDescription` to `StyleColor.rgbaString`.
- Changed `StyleColor`'s `Encodable` implementation to always encode an rgba color string instead of encoding an rgba expression
- Updated the extension on `UIColor` that adds `ExpressionArgumentConvertible` to return an rgba color string instead of an rgba expression.
- Style properties on annotation managers are now synced to the layer each display link just like their annotations. Use `syncSourceAndLayerIfNeeded()` (formerly named `syncAnnotationsIfNeeded()`) to force the sync to happen earlier.
- The `layerType` argument to `Style._layerPropertyDefaultValue(for:property:)` is now of type `LayerType` instead of `String`
- `Expression` decoding will now fail if the operator is missing
- `PointAnnotationManager.textVariableAnchor` is now of type `[TextAnchor]?` instead of `[String]?`
- `PointAnnotationManager.textWritingMode` is now of type `[TextWritingMode]?` instead of `[String]?`

#### Bug Fixes

- Updated annotations to use `rgbaString` and `init(rgbaString:)` when serializing and deserializing `StyleColor`s (Fixes https://github.com/mapbox/mapbox-maps-ios/issues/636)
- Annotation managers now properly restore the default values of any annotation or common style properties that are reset to nil.
- Fixed Expression decoding when second array element could be an operator

#### Additions

- Added a public, failable, component-wise initializer to `StyleColor`
- Updated `StyleColor`'s `Decodable` support to be able to handle rgba color strings as well as rgba expressions (Fixes https://github.com/mapbox/mapbox-maps-ios/issues/577)
- Made generated enums conform to `CaseIterable`

#### Other Changes

- `StyleColor` now stores its data as component-wise `Double`s instead of as an expression.
- Added an internal, failable, expression-based initializer to `StyleColor`
- Added an internal, failable, rgba-string-based initializer to `StyleColor`
- Updated the codebase to use type inference with `UIColor` when possible
- Moved the extension on `UIColor` that adds `ExpressionArgumentConvertible` conformance to ExpressionArgumentBuilder.swift
- Update annotation managers to store common layer styles in a dictionary similarly to how annotations store their styles. This provides better handling of `nil` since the `value` parameter of `StyleManager.setStyleLayerPropertyForLayerId(_:property:value:)` is marked `nonnull` in Objective-C. The previous implementation silenced a warning about this by casting the optional to Any.
- Updated style and annotation templates to consistently include a trailing newline
- Made generated test classes final
- Allow linter to run on generated code, and fix lint
- Added integration tests for syncing each annotation and annotation manager property to the layer
- Added Random.swift test helpers

### CODE REVIEW

**Comment by nishant-karajgikar** [comment]
Could we run device tests on this PR?

File: Sources/MapboxMaps/Style/Types/StyleColor.swift [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    // MARK: - RGBA String
+
+    /// Creates a `StyleColor` from an rgba color string as defined
+    /// in the Mapbox Style Spec. Returns nil if the string is not a valid
+    /// rgba color string.
+    /// - Parameter rgbaString: An rgba color string
+    internal init?(rgbaString: String) {
+        let nsString = NSString(string: rgbaString)
+        let regex = try! NSRegularExpression(pattern: "rgba\\( *([0-9.]+) *, *([0-9.]+) *, *([0-9.]+) *, *([0-9.]+) *\\)", options: [])
```
Comment (macdrevx) [inline Sources/MapboxMaps/Style/Types/StyleColor.swift:84]: should clamp to the beginning and end of the input string, allowing leading and trailing whitespace

File: Tests/MapboxMapsTests/Style/StyleColorTests.swift [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    var rgbaString: String {
+        "rgba(\(red!), \(green!), \(blue!), \(alpha!))"
+    }
+
+    override func setUp() {
+        super.setUp()
+        red = Double.random(in: 0...255)
+        green = Double.random(in: 0...255)
+        blue = Double.random(in: 0...255)
+        alpha = Double.random(in: 0...1)
```
Comment (macdrevx) [inline Tests/MapboxMapsTests/Style/StyleColorTests.swift:20]: Use type inference

File: Tests/MapboxMapsTests/Style/StyleColorTests.swift [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            blue!
+            alpha!
+        }
+
+        verify(StyleColor(expression: expression))
+    }
+
+    func testExpressionInitFailureWrongOperator() {
+        let op = Expression.Operator.allCases.filter { $0 != .rgba }.randomElement()!
+        let expression = Exp(op) {}
```
Comment (macdrevx) [inline Tests/MapboxMapsTests/Style/StyleColorTests.swift:85]: Add a set of valid arguments to show that the failure is due to the operator

File: Tests/MapboxMapsTests/Style/StyleColorTests.swift [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    func testRGBAStringInitFailureInvalidStructure() {
+        XCTAssertNil(StyleColor(rgbaString: ""))
+        XCTAssertNil(StyleColor(rgbaString: "rgba (0, 0, 0, 0)"))
+        XCTAssertNil(StyleColor(rgbaString: "rgb(0, 0, 0)"))
+        XCTAssertNil(StyleColor(rgbaString: "rgb(0, 0, 0, 0)"))
+        XCTAssertNil(StyleColor(rgbaString: "rgba(0, 0, 0)"))
+        XCTAssertNil(StyleColor(rgbaString: "rgba(a, b, c)"))
+        XCTAssertNil(StyleColor(rgbaString: "rgba(0.0.0, 0.0.0, 0.0.0, 0.0.0)"))
+    }
```
Comment (macdrevx) [inline Tests/MapboxMapsTests/Style/StyleColorTests.swift:135]: add test case for random leading or trailing characters

**Comment by macdrevx** [comment]
Disable line gradient

**Review by nishant-karajgikar — APPROVED** [review]

**Pull request merged**